// @statsig/session-replay/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.forceStartRecording = exports.stopRecording = exports.startRecording = exports.StatsigTriggeredSessionReplayPlugin = exports.StatsigSessionReplayPlugin = exports.runStatsigTriggeredSessionReplay = exports.runStatsigSessionReplay = exports.SessionReplay = exports.SessionReplayClient = void 0;
const client_core_1 = require("@statsig/client-core");
const SessionReplay_1 = require("./SessionReplay");
Object.defineProperty(exports, "SessionReplay", { enumerable: true, get: function () { return SessionReplay_1.SessionReplay; } });
Object.defineProperty(exports, "StatsigSessionReplayPlugin", { enumerable: true, get: function () { return SessionReplay_1.StatsigSessionReplayPlugin; } });
Object.defineProperty(exports, "runStatsigSessionReplay", { enumerable: true, get: function () { return SessionReplay_1.runStatsigSessionReplay; } });
const SessionReplayBase_1 = require("./SessionReplayBase");
Object.defineProperty(exports, "forceStartRecording", { enumerable: true, get: function () { return SessionReplayBase_1.forceStartRecording; } });
Object.defineProperty(exports, "startRecording", { enumerable: true, get: function () { return SessionReplayBase_1.startRecording; } });
Object.defineProperty(exports, "stopRecording", { enumerable: true, get: function () { return SessionReplayBase_1.stopRecording; } });
const SessionReplayClient_1 = require("./SessionReplayClient");
Object.defineProperty(exports, "SessionReplayClient", { enumerable: true, get: function () { return SessionReplayClient_1.SessionReplayClient; } });
const TriggeredSessionReplay_1 = require("./TriggeredSessionReplay");
Object.defineProperty(exports, "StatsigTriggeredSessionReplayPlugin", { enumerable: true, get: function () { return TriggeredSessionReplay_1.StatsigTriggeredSessionReplayPlugin; } });
Object.defineProperty(exports, "runStatsigTriggeredSessionReplay", { enumerable: true, get: function () { return TriggeredSessionReplay_1.runStatsigTriggeredSessionReplay; } });
Object.assign((0, client_core_1._getStatsigGlobal)(), {
    SessionReplayClient: SessionReplayClient_1.SessionReplayClient,
    SessionReplay: SessionReplay_1.SessionReplay,
    runStatsigSessionReplay: SessionReplay_1.runStatsigSessionReplay,
    runStatsigTriggeredSessionReplay: TriggeredSessionReplay_1.runStatsigTriggeredSessionReplay,
    StatsigSessionReplayPlugin: SessionReplay_1.StatsigSessionReplayPlugin,
    StatsigTriggeredSessionReplayPlugin: TriggeredSessionReplay_1.StatsigTriggeredSessionReplayPlugin,
    startRecording: SessionReplayBase_1.startRecording,
    stopRecording: SessionReplayBase_1.stopRecording,
    forceStartRecording: SessionReplayBase_1.forceStartRecording,
});
