// @statsig/session-replay/src/SessionReplayUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNewRRWebConfigWithPrivacySettings = exports._appendSlicedMetadata = exports._slicePayload = exports._makeLoggableRrwebEvent = exports.MAX_LOGS = exports.MAX_INDIVIDUAL_EVENT_BYTES = exports.REPLAY_ENQUEUE_TRIGGER_BYTES = void 0;
const client_core_1 = require("@statsig/client-core");
const REPLAY_SLICE_BYTES = 1024 * 1024; // 1 MB
exports.REPLAY_ENQUEUE_TRIGGER_BYTES = 1024 * 10; // 10 KB
exports.MAX_INDIVIDUAL_EVENT_BYTES = 1024 * 1024 * 10; // 10 MB
exports.MAX_LOGS = 10000; // 10K logs
function _makeLoggableRrwebEvent(slice, payload, sessionID, data, sdkInstanceID) {
    const metadata = {
        session_start_ts: String(data.startTime),
        session_end_ts: String(data.endTime),
        clicks_captured_cumulative: String(data.clickCount),
        rrweb_events: slice,
        rrweb_payload_size: String(payload.length),
        session_replay_sdk_version: client_core_1.SDK_VERSION,
        sdk_instance_id: sdkInstanceID,
    };
    return {
        eventName: 'statsig::session_recording',
        value: sessionID,
        metadata,
    };
}
exports._makeLoggableRrwebEvent = _makeLoggableRrwebEvent;
function _slicePayload(payload) {
    const parts = [];
    for (let i = 0; i < payload.length; i += REPLAY_SLICE_BYTES) {
        parts.push(payload.slice(i, i + REPLAY_SLICE_BYTES));
    }
    return parts;
}
exports._slicePayload = _slicePayload;
function _appendSlicedMetadata(metadata, slicedID, sliceIndex, sliceCount, sliceByteSize) {
    metadata.sliced_id = slicedID;
    metadata.slice_index = String(sliceIndex);
    metadata.slice_count = String(sliceCount);
    metadata.slice_byte_size = String(sliceByteSize);
}
exports._appendSlicedMetadata = _appendSlicedMetadata;
function getNewRRWebConfigWithPrivacySettings(originalOptions, privacySettings) {
    const maskValue = (value) => {
        return value.replace(/./g, '*');
    };
    function getNearestPrivacyMatch(element, maskedSelectors = [], unmaskedSelectors = []) {
        let current = element;
        while (current !== null) {
            if (maskedSelectors.some((sel) => current === null || current === void 0 ? void 0 : current.matches(sel))) {
                return 'masked';
            }
            if (unmaskedSelectors.some((sel) => current === null || current === void 0 ? void 0 : current.matches(sel))) {
                return 'unmasked';
            }
            current = current.parentElement;
        }
        return null;
    }
    const blockSelector = privacySettings.blocked_elements
        ? privacySettings.blocked_elements.join(', ')
        : undefined;
    const maskTextFn = (value, element) => {
        if (!element) {
            return privacySettings.privacy_mode === 'max' ? maskValue(value) : value;
        }
        const nearest = getNearestPrivacyMatch(element, privacySettings.masked_elements, privacySettings.unmasked_elements);
        if (nearest === 'masked') {
            return maskValue(value);
        }
        if (nearest === 'unmasked') {
            return value;
        }
        return privacySettings.privacy_mode === 'max' ? maskValue(value) : value;
    };
    const maskInputFn = (value, element) => {
        var _a, _b;
        if ((_a = privacySettings.masked_elements) === null || _a === void 0 ? void 0 : _a.some((sel) => element.closest(sel))) {
            return maskValue(value);
        }
        if ((_b = privacySettings.unmasked_elements) === null || _b === void 0 ? void 0 : _b.some((sel) => element.closest(sel))) {
            return value;
        }
        return privacySettings.privacy_mode === 'max' ||
            privacySettings.privacy_mode === 'input'
            ? maskValue(value)
            : value;
    };
    return Object.assign(Object.assign({}, originalOptions), { maskTextFn: maskTextFn, maskInputFn, maskTextSelector: '*', maskAllInputs: true, maskInputOptions: undefined, blockSelector });
}
exports.getNewRRWebConfigWithPrivacySettings = getNewRRWebConfigWithPrivacySettings;
